Update content model for customizable select

josepharhar opened the pull request:
This PR updates the content model for the `<select>`, `<option>`, and `<optgroup>` elements in support of customizable `<select>`.

The major additions are:
- `<select>` and `<optgroup>` can contain `<div>` elements.
- `<option>` elements can contain `<div>` and non-interactive phrasing content.
- `<optgroup>` can have a child `<legend>` to represent its label.

Fixes https://github.com/whatwg/html/issues/10317

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
When editing this comment after the PR is created, check that PR-Preview doesn't overwrite your changes.
If you think your PR is ready to land, please double-check that the build is passing and the checklist is complete before pinging.
-->

- [x] At least two implementers are interested (and none opposed):
   * Chrome
   * Mozilla: https://github.com/mozilla/standards-positions/issues/1060
   * WebKit: https://github.com/WebKit/standards-positions/issues/386
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://wpt.fyi/results/html/semantics/forms/the-select-element/customizable-select/
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: [Already implemented](https://issues.chromium.org/issues/40146374)
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1944403
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=286642
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/content/issues/37857
- [x] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/10586/dom.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">/dom.html</a>  ( <a href="https://whatpr.org/html/10586/2e87c0f...26edf65/dom.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">diff</a> )
<a href="https://whatpr.org/html/10586/form-elements.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">/form-elements.html</a>  ( <a href="https://whatpr.org/html/10586/2e87c0f...26edf65/form-elements.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">diff</a> )
<a href="https://whatpr.org/html/10586/grouping-content.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">/grouping-content.html</a>  ( <a href="https://whatpr.org/html/10586/2e87c0f...26edf65/grouping-content.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">diff</a> )
<a href="https://whatpr.org/html/10586/index.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">/index.html</a>  ( <a href="https://whatpr.org/html/10586/2e87c0f...26edf65/index.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">diff</a> )
<a href="https://whatpr.org/html/10586/indices.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">/indices.html</a>  ( <a href="https://whatpr.org/html/10586/2e87c0f...26edf65/indices.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">diff</a> )
<a href="https://whatpr.org/html/10586/rendering.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">/rendering.html</a>  ( <a href="https://whatpr.org/html/10586/2e87c0f...26edf65/rendering.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">diff</a> )
<a href="https://whatpr.org/html/10586/scripting.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">/scripting.html</a>  ( <a href="https://whatpr.org/html/10586/2e87c0f...26edf65/scripting.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">diff</a> )
<a href="https://whatpr.org/html/10586/text-level-semantics.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">/text-level-semantics.html</a>  ( <a href="https://whatpr.org/html/10586/2e87c0f...26edf65/text-level-semantics.html" title="Last updated on May 13, 2025, 6:20 PM UTC (26edf65)">diff</a> )

zcorpan commented:
Is the idea to allow `<select><div><h1><option>` and such? Or only `<select><div><div><option>`?

If the former, you need to change the definition of phrasing content to include "`option` (if it is a descendant of a `select` element), `optgroup` (if it is a descendant of a `select` element)".

If the latter, you need to change the content model if `div` and `span` to switch on whether there's an ancestor `select` element, and if so allow "select element inner content elements". (But still disallow nesting of `optgroup` and `option`)

zcorpan commented on source:
```diff
@@ -12785,6 +12785,34 @@ https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C%21DOCTYPE%20HTML%3E%
+
+   <li><code>span</code> elements.</li>
+
+   <li><code>img</code> elements.</li>
```
Is the idea to allow `img` *between* options?

scottaohara replied:
i think people will generally just want images within their options.  they'll probably also sometimes want to use them within the `legend` for `optgroup`s.  If it were just up to me, i'd limit their use to being children of these elements.

per discussions that were had on this, the primary use case people seemed to think authors might want them for, outside of options, would be for decorative reasons.  but i don't see why that couldn't be done with the allowed divs/spans and CSS.

josepharhar commented:
> Is the idea to allow `<select><div><h1><option>` and such? Or only `<select><div><div><option>`?
> 
> If the former, you need to change the definition of phrasing content to include "`option` (if it is a descendant of a `select` element), `optgroup` (if it is a descendant of a `select` element)".
> 
> If the latter, you need to change the content model if `div` and `span` to switch on whether there's an ancestor `select` element, and if so allow "select element inner content elements". (But still disallow nesting of `optgroup` and `option`)

h1 elements are not supposed to be part of the content model, so I implemented the div and span switching by pulling out the content models for option and optgroup and using them in div and span.

josepharhar commented on source:
```diff
@@ -53203,7 +53231,8 @@ interface <dfn interface>HTMLButtonElement</dfn> : <span>HTMLElement</span> {
    <dd>Where <span>phrasing content</span> is expected.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd>Zero or more <code>option</code>, <code>optgroup</code>, <code>hr</code>, and <span data-x="script-supporting elements">script-supporting</span> elements.</dd>
+   <dd>Zero or more <span>select element inner content elements</span>.</dd>
+   <dd>Zero or one child <code>button</code>.</dd>
```
The button should probably go first, and I wanted to keep the "zero or one" concept with the button, so I wrote it like this:
```
<dd>Zero or one <code>button</code> element followed by zero or more <span>select element inner content elements</span>.</dd>
```
Look good?

josepharhar commented on source:
```diff
@@ -53932,9 +53961,10 @@ interface <dfn interface>HTMLDataListElement</dfn> : <span>HTMLElement</span> {
    <dt><span data-x="concept-element-categories">Categories</span>:</dt>
    <dd>None.</dd>
```
added

josepharhar commented on source:
```diff
@@ -53932,9 +53961,10 @@ interface <dfn interface>HTMLDataListElement</dfn> : <span>HTMLElement</span> {
+   <dd>As a descendant of a <code>select</code> element.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd>Zero or more <code>option</code> and <span data-x="script-supporting elements">script-supporting</span> elements.</dd>
+   <dd>Zero or more <span>select element inner content elements</span>.</dd>
```
I've implemented this in the new "optgroup inner element content elements" category

josepharhar commented on source:
```diff
@@ -53932,9 +53961,10 @@ interface <dfn interface>HTMLDataListElement</dfn> : <span>HTMLElement</span> {
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd>Zero or more <code>option</code> and <span data-x="script-supporting elements">script-supporting</span> elements.</dd>
+   <dd>Zero or more <span>select element inner content elements</span>.</dd>
+   <dd>Zero or one <code>legend</code> elements.</dd>
```
done

josepharhar commented on source:
```diff
@@ -54029,17 +54059,15 @@ interface <dfn interface>HTMLOptGroupElement</dfn> : <span>HTMLElement</span> {
-   <dd>If the element has no <code data-x="attr-option-label">label</code> attribute and is a child
-   of a <code>datalist</code> element: <span data-x="text content">Text</span>.</dd>
+   <dd>If the element has no <code data-x="attr-option-label">label</code> attribute: <span>Text content</span> and zero or more <code>div</code>, <code>span</code>, <code>noscript</code>, <code>img</code>, <span>SVG <code>svg</code></span>, and <span data-x="script-supporting elements">script-supporting</span> elements.</dd>
+   <dd><div class="note">If an <code>img</code> is used within an <code>option</code> and there is no other text within the <code>option</code>, then the <code>option</code> won't have an accessible name unless the <code>img</code> also has a non-empty <code data-x="attr-img-alt">alt</code> attribute.</div></dd>
```
I thought this was what @scottaohara suggested, but as per [this comment](https://github.com/whatwg/html/issues/10317#issuecomment-2318219319), it sounds like we can add notes and examples as a follow up. I removed the note for now.

scottaohara replied:
besides the placement could you elaborate on what was wrong with the note @zcorpan?  i might have phrased it a bit differently, but it seemed to be saying `<option><img src=...></option>` would have no accessible name, and that is correct.

zcorpan replied:
@scottaohara The `option` element or the `img` element could have `aria-label` or `aria-labelledby` attributes.

mtrootyy commented on source:
```diff
@@ -63298,6 +63372,9 @@ not-slash     = %x0000-002E / %x0030-10FFFF
    <dd><span>Phrasing content</span>.</dd>
+   <dd><span>Select element inner content elements</span>.</dd>
+   <dd><span>Option element inner content elements</span>.</dd>
+   <dd><span>Optgroup element inner content elements</span>.</dd>
```
In "Content model:" of the noscript element,the optgroup element and the option element should be allowed in the noscript element which is a descendant of a select element.

josepharhar commented on source:
```diff
@@ -19774,6 +19836,8 @@ and is further discussed below.&lt;/div></code></pre>
    <dt><span data-x="concept-element-categories">Categories</span>:</dt>
    <dd><span>Flow content</span>.</dd>
+   <dd><span>Select element inner content elements</span>.</dd>
+   <dd><span>Optgroup element inner content elements</span>.</dd>
```
Done, thanks!

josepharhar commented on source:
```diff
@@ -21508,12 +21572,17 @@ included with Exhibit B.
    <dd><span>Flow content</span>.</dd>
    <dd><span>Palpable content</span>.</dd>
+   <dd><span>Select element inner content elements</span>.</dd>
+   <dd><span>Option element inner content elements</span>.</dd>
```
Done, thanks!

josepharhar commented on source:
```diff
@@ -21508,12 +21572,17 @@ included with Exhibit B.
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
    <dd>If the element is a child of a <code>dl</code> element: one or more <code>dt</code> elements followed by one or more <code>dd</code> elements, optionally intermixed with <span>script-supporting elements</span>.</dd>
-   <dd>If the element is not a child of a <code>dl</code> element: <span>flow content</span>.</dd>
+   <dd>Otherwise, if the element is a descendant of an <code>option</code> element: Zero or more <span>option element inner content elements</span>.</dd>
```
done, thanks!

josepharhar commented on source:
```diff
@@ -24325,10 +24394,15 @@ wormhole connection.&lt;/mark>&lt;/p></code></pre>
    <dd><span>Flow content</span>.</dd>
    <dd><span>Phrasing content</span>.</dd>
    <dd><span>Palpable content</span>.</dd>
+   <dd><span>Select element inner content elements</span>.</dd>
```
done, thanks!

josepharhar commented on source:
```diff
@@ -24325,10 +24394,15 @@ wormhole connection.&lt;/mark>&lt;/p></code></pre>
    <dd><span>Palpable content</span>.</dd>
+   <dd><span>Select element inner content elements</span>.</dd>
+   <dd><span>Option element inner content elements</span>.</dd>
+   <dd><span>Optgroup element inner content elements</span>.</dd>
```
done

josepharhar commented on source:
```diff
@@ -24325,10 +24394,15 @@ wormhole connection.&lt;/mark>&lt;/p></code></pre>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd><span>Phrasing content</span>.</dd>
+   <dd>If the element is a descendant of an <code>option</code> element: Zero or more <span>option element inner content elements</span>.</dd>
+   <dd>Otherwise, if the element is a descendant of an <code>optgroup</code> element: Zero or more <span>optgroup element inner content elements</span>.</dd>
```
done

josepharhar commented on source:
```diff
@@ -29053,6 +29127,9 @@ interface <dfn interface>HTMLSourceElement</dfn> : <span>HTMLElement</span> {
    <dd><span>Form-associated element</span>.</dd>
    <dd>If the element has a <code data-x="attr-hyperlink-usemap">usemap</code> attribute: <span>Interactive content</span>.</dd>
    <dd><span>Palpable content</span>.</dd>
+   <dd><span>Select element inner content elements</span>.</dd>
```
done

josepharhar commented on source:
```diff
@@ -29053,6 +29127,9 @@ interface <dfn interface>HTMLSourceElement</dfn> : <span>HTMLElement</span> {
    <dd><span>Palpable content</span>.</dd>
+   <dd><span>Select element inner content elements</span>.</dd>
+   <dd><span>Option element inner content elements</span>.</dd>
+   <dd><span>Optgroup element inner content elements</span>.</dd>
```
done

josepharhar commented on source:
```diff
@@ -12785,6 +12785,68 @@ https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C%21DOCTYPE%20HTML%3E%
+
+   <li><code>img</code> elements.</li>
+
+   <li><span>SVG <code>svg</code></span> elements.</li>
```
done

josepharhar commented on source:
```diff
@@ -21508,12 +21572,17 @@ included with Exhibit B.
    <dd>If the element is a child of a <code>dl</code> element: one or more <code>dt</code> elements followed by one or more <code>dd</code> elements, optionally intermixed with <span>script-supporting elements</span>.</dd>
-   <dd>If the element is not a child of a <code>dl</code> element: <span>flow content</span>.</dd>
+   <dd>Otherwise, if the element is a descendant of an <code>option</code> element: Zero or more <span>option element inner content elements</span>.</dd>
+   <dd>Otherwise, if the element is a descendant of an <code>optgroup</code> element: Zero or more <span>optgroup element inner content elements</span>.</dd>
```
done

josepharhar commented on source:
```diff
@@ -53930,11 +54007,11 @@ interface <dfn interface>HTMLDataListElement</dfn> : <span>HTMLElement</span> {
+   <dd>As a descendant of a <code>select</code> element.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd>Zero or more <code>option</code> and <span data-x="script-supporting elements">script-supporting</span> elements.</dd>
+   <dd>Zero or one <code>legend</code> element followed by zero or more <span>optgroup element inner content elements</span>.</dd>
```
done

josepharhar commented on source:
```diff
@@ -53203,7 +53280,7 @@ interface <dfn interface>HTMLButtonElement</dfn> : <span>HTMLElement</span> {
    <dd>Where <span>phrasing content</span> is expected.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd>Zero or more <code>option</code>, <code>optgroup</code>, <code>hr</code>, and <span data-x="script-supporting elements">script-supporting</span> elements.</dd>
+   <dd>Zero or one <code>button</code> element followed by zero or more <span>select element inner content elements</span>.</dd>
```
I included this concept in the existing sentence

mtrootyy commented on source:
```diff
@@ -55773,6 +55842,7 @@ interface <dfn interface>HTMLFieldSetElement</dfn> : <span>HTMLElement</span> {
    <dd>As the <span>first child</span> of a <code>fieldset</code> element.</dd>
+   <dd>As the <span>first child</span> of an <code>optgroup</code> element.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
    <dd><span>Phrasing content</span>, optionally intermixed with <span>heading content</span>.</dd>
```
When the legend element is first child of an optgroup element, "heading content" doesn't need to be allowed.

josepharhar replied:
I appended that casing to the end of this sentence

josepharhar commented on source:
```diff
@@ -53203,7 +53280,7 @@ interface <dfn interface>HTMLButtonElement</dfn> : <span>HTMLElement</span> {
    <dd>Where <span>phrasing content</span> is expected.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd>Zero or more <code>option</code>, <code>optgroup</code>, <code>hr</code>, and <span data-x="script-supporting elements">script-supporting</span> elements.</dd>
+   <dd>Zero or one <code>button</code> element followed by zero or more <span>select element inner content elements</span>.</dd>
```
> Description of "As first child of select element" should be added to "Contexts in which this element can be used:" of section of "4.10.6 The button element".

Done
 
> In this case, this button element is neither of "Submit Button" state, "Reset Button" state or "Button" state. Therefore, a fourth state of button element sholud be defined.
> 
> In addition, Shouldn't The following sentent be added to section of "4.10.6 The button element"?
> 
> ```
> If the button element is a fourth state, "popovertarget", "popovertargetaction", "type", "name" and "value" attribute are must not be spedified.
> ```

Those states are based on the type attribute. Since we didn't add any states when adding popovertarget, I don't think we should add a state here.

I suppose that we could say that the type attribute should not be set when the button is the first child of a select or when it has the popovertarget attribute though? Perhaps as a follow-up?

mtrootyy replied:
In "type" attribute of button element, the attribute's missing value default and invalid value default are both the Submit Button state.
Therefore, if "type" attribute would be simply prohibited, button element which is child of select element would be Submit Button state.
Because this is not appropriate, shouldn't the fourth state of button element be defined?

scottaohara replied:
defining a fourth button type sounds like it could be potential scope creep, but @mtrootyy does make a good point that it is probably worth noting that this button is _not_ a submit button, and that it also probably shouldn't even allow many of the attributes that would be valid for a button in other contexts.

josepharhar replied:
Since the button is now inert, I don't think we need to put extra special casing in to undo what buttons do in other places. This was part of the motivation of using inert, so that we could remove the logic to check whether the button is in a select when doing form submission from a submit button.

josepharhar replied:
I added a note about this: https://github.com/whatwg/html/pull/10586/commits/473ca72a08e0e6fb52af1373cb8a66b0ca51e265

josepharhar commented on source:
```diff
@@ -12785,6 +12785,68 @@ https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C%21DOCTYPE%20HTML%3E%
+  <ul class="brief category-list">
+   <li><span>Text content</span>.</li>
+
+   <li><span>Palpable content</span> except for <span>interactive content</span>.</li>
```
Sounds reasonable, done. Thanks!

mtrootyy commented on source:
```diff
@@ -54027,19 +54099,16 @@ interface <dfn interface>HTMLOptGroupElement</dfn> : <span>HTMLElement</span> {
-   <dd>If the element has no <code data-x="attr-option-label">label</code> attribute and is a child
-   of a <code>datalist</code> element: <span data-x="text content">Text</span>.</dd>
+   <dd>If the element has no <code data-x="attr-option-label">label</code> attribute: Zero or more <span>option element inner content elements</span>.</dd>
    <dt><span data-x="concept-element-attributes">Content attributes</span>:</dt>
```
"Content model:" should become the following.
```
    If the element has a label attribute and a value attribute: Nothing.
    If the element has a label: Text.
    If the element has no value attribute: Text.
    If the element has no label attribute and has a value attribute: Zero or more option element inner content elements.
```
Because the submitted value must be simply text, and "option element inner content elements" is only used to decorate label text.

josepharhar replied:
I'm not sure I want to force authors to set a value attribute for simple cases like this:
```html
<select>
  <option><svg>...</svg><span>Purple</span></option>
  <option><svg>...</svg><span>Green</span></option>
</select>
```
In this case, we can just use the text contents of the option, which would be "purple" and "green", just like we already do without the value attribute.

I suppose it would make sense to enforce that there is either a value attribute or some text node descendant of the option. Do you agree?

mtrootyy replied:
```
<select name="a">
<option value="2^2"><span>2<sup>2</sup></span></option>
<option value="square root of 3"><math><msqrt><mn>3</mn></msqrt></math></option>
</select>
```

In the above case, if there is no value attribute, the submitted value will be "22","3".
Because there are many cases like this, shouldn't it be enforce to specify the value attribute?

josepharhar replied:
If the author is actually using the value property of the select for anything, then I think they will notice odd cases like this and use the value attribute, just like they are already doing today since it is currently optional.

I could imagine cases today where the author puts non-ascii unicode characters into the option's text, and then has difficulties parsing it later when its submitted to their server in a form. I don't want to start requiring an explicit value attribute in cases like this either.

mtrootyy replied:
If value attribute shouldn't be required in this case, shouldn't some kind of note regarding above points be described?

josepharhar replied:
All I could imagine the note saying is that when there is no value attribute, the option's text content is used to create a value, which is already described here: https://html.spec.whatwg.org/#attr-option-value

josepharhar replied:
I added a note about this: https://github.com/whatwg/html/pull/10586/commits/473ca72a08e0e6fb52af1373cb8a66b0ca51e265

annevk commented on source:
```diff
@@ -12785,6 +12785,68 @@ https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C%21DOCTYPE%20HTML%3E%
+  <h6>Select element inner content elements</h6>
+
+  <p><dfn>Select element inner content elements</dfn> are <code>option</code> elements and other
+  elements which are used to group and decorate them with a <code>select</code> element.</p>
```
This reads weird.

josepharhar replied:
I re-wrote it like this: ... are the elements which are allowed as descendants of select elements.

josepharhar replied:
I'm going to guess that this looks ok now

annevk commented on source:
```diff
@@ -12785,6 +12785,68 @@ https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C%21DOCTYPE%20HTML%3E%
+
+   <li><code>noscript</code> elements.</li>
+
+   <li><code>div</code> elements.</li>
```
Why `div` but not `span`?

Also: I recommend looking at this rendered. This does not look good and doesn't match the other lists of this nature.

josepharhar replied:
I made select only have divs and options only have spans in response to a meeting comment from @fantasai. However, since then, @brechtdr commented [here](https://github.com/whatwg/html/issues/10317#issuecomment-2417419742) in support of allowing both div and span, and Scott sounded supportive (his original content model allowed for both). Since it sounds like you also think we should do both div and span, I will change it back to allow both of them.

josepharhar replied:
> Also: I recommend looking at this rendered. This does not look good and doesn't match the other lists of this nature.

Thanks, I missed that and fixed it up

josepharhar replied:
I re-added div and span

annevk commented on source:
```diff
@@ -12785,6 +12785,68 @@ https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C%21DOCTYPE%20HTML%3E%
   </ul>
 
 
+  <h6>Select element inner content elements</h6>
```
I think it's rather confusing to have both `select` element and "Select element".

josepharhar replied:
I'm not sure if I should reference `<code>select</code>` in the `<dfn>` for this concept, and I didn't make all of the references to it do that either.

Should I do that instead? Or are there other places where I'm using `<code>select</code>` that should just be "select"?

zcorpan replied:
You can use a `code` element in a term without it itself being a reference (with `data-x=""`). Example: https://html.spec.whatwg.org/#the-head-element-2

josepharhar replied:
I used `<code>select</code>` everywhere, and did it for option and optgroup too

scottaohara commented:
we need to update the optgroup element with it's allowance of the `legend` element.

similar to option where someone can either use text, or label/value attributes - the optgroup should either have a label for the group rendered by the use of the `label` attribute OR by someone using a `legend` element.  

If both are defined (label attribute and legend element) one should win out over the other, like what happens with option and the current select > option.

scottaohara commented on source:
```diff
@@ -55773,8 +55842,9 @@ interface <dfn interface>HTMLFieldSetElement</dfn> : <span>HTMLElement</span> {
+   <dd>As the <span>first child</span> of an <code>optgroup</code> element.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd><span>Phrasing content</span>, optionally intermixed with <span>heading content</span>.</dd>
+   <dd><span>Phrasing content</span>, optionally intermixed with <span>heading content</span> if this is not the <span>first child</span> of an <code>optgroup</code> element.</dd>
```
it might make more sense to split this out into two lines.  the line as it is in the spec for a legend as a child of a fieldset

and a separate line to indicate that in the context of an optgroup, a legend may contain phrasing content but there must be no [interactive content](https://whatpr.org/html/10586/dom.html#interactive-content-2) descendant and no descendant with the [tabindex](https://whatpr.org/html/10586/interaction.html#attr-tabindex) attribute specified.

josepharhar replied:
Agreed, done!

scottaohara commented:
related to my [above comment](https://github.com/whatwg/html/pull/10586#issuecomment-2429931537), this text from the `optgroup` element needs to be updated, since `label` isn't the only way to label an `optgroup` anymore

>The label attribute must be specified. Its value gives the name of the group, for the purposes of the user interface. User agents should use this attribute's value when labeling the group of [option](https://whatpr.org/html/10586/form-elements.html#the-option-element) elements in a [select](https://whatpr.org/html/10586/form-elements.html#the-select-element) element.


Probably should read something like the following (bolded text being my additions / changes)


>**If the `optgroup` does not contain a `legend` element,** the label attribute must be specified. Its value gives the name of the group, for the purposes of the user interface. User agents should use this attribute's value when labeling the group of [option](https://whatpr.org/html/10586/form-elements.html#the-option-element) elements in a [select](https://whatpr.org/html/10586/form-elements.html#the-select-element) element.

Re: "User agents should use this attribute's value..." I'm actually surprised that wasn't a UA 'must'.  But with the addition of the legend element, this seems like it could be the place to clarify what takes priority if both are declared by the author.

mtrootyy commented:
@scottaohara
To ensure backwards compatible, even if the optgroup element contains a legend element, the label attribute on the optgroup element must be specified.

scottaohara commented:
@mtrootyy then even more reason the spec needs to be updated to mention this, and that work be done to not render both the text from the label attribute AND the legend - which is what's happening now and seems rather rubbish.

mtrootyy commented on source:
```diff
@@ -55773,8 +55842,9 @@ interface <dfn interface>HTMLFieldSetElement</dfn> : <span>HTMLElement</span> {
    <dd>None.</dd>
    <dt><span data-x="concept-element-contexts">Contexts in which this element can be used</span>:</dt>
    <dd>As the <span>first child</span> of a <code>fieldset</code> element.</dd>
+   <dd>As the <span>first child</span> of an <code>optgroup</code> element.</dd>
```
Some sentence should be added to prose of section of "4.10.16 The legend element".
For example;
```
The legend element represents a caption for the rest of the contents of the legend element's parent fieldset element, if any.
```
　↓

```
The legend element represents a caption for the rest of the contents of the legend element's parent fieldset element, if any.
Or, the legend element represents label of it's parent optgroup element.
```

josepharhar replied:
Done, thanks!

josepharhar commented:
> This could use some examples. Or are we adding those as part of another PR?

I tried adding an example initially, but I put it in the wrong spot and/or otherwise messed it up: https://github.com/whatwg/html/pull/10586#discussion_r1761323296

Scott suggested doing a follow up, so that's my plan: https://github.com/whatwg/html/issues/10317#issuecomment-2318219319

mtrootyy commented on source:
```diff
@@ -24529,10 +24604,16 @@ wormhole connection.&lt;/mark>&lt;/p></code></pre>
    <dd>Where <span>phrasing content</span> is expected.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd><span>Phrasing content</span>.</dd>
+   <dd>If the element is a descendant of a <code>select</code> element: Zero or more <span>select element inner content elements</span>.</dd>
```
If we did that, the span element is allowed to contain the div element.
Is that really okay?

josepharhar replied:
I think it is ok.

It was pointed out earlier that options shouldn't contain divs, but then we got feedback that we should allow either of them. Similarly, I think that if we make rules here about combining divs and spans it won't go over well.

josepharhar replied:
Based on the thread with scott, I am removing span from select/optgroup and only allowing them in options again, so I fixed this

scottaohara commented on source:
```diff
@@ -12975,6 +12975,74 @@ https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C%21DOCTYPE%20HTML%3E%
+
+   <li><code>div</code></li>
+
+   <li><code>span</code></li>
```
the adjustments to the content model and the back and forth about span allowances have maybe made span show up in more places than it needs to be - leading to [this good catch](https://github.com/whatwg/html/pull/10586#pullrequestreview-2400186705) from @mtrootyy

A span should be allowed for use for the elements that allow phrasing content as part of their content model - and by just referencing the phrasing content allowance, the spec doesn't need to also call out spans then.  Otherwise, a div should be sufficient for elements like select and optgroup. 



in general i'm wondering if these sub categorizations of inner select element content models is necessary, especially if it's just the elements themselves and the div element that need to reference the allowed children (if span is updated to only be used within an option - by being included in the category of phrasing content).

And for most of the cases where div is being allowed, couldn't it be identified as having a transparent content model?
re:
>The content model of a [transparent](https://whatpr.org/html/10586/dom.html#transparent) element is derived from the content model of its parent element: the elements required in the part of the content model that is "transparent" are the same elements as required in the part of the content model of the parent of the transparent element in which the transparent element finds itself.

maybe that doesn't work for optgroup since it needs to start with a legend - but why not for select or option elements?  Generally curious / want to understand what i might be missing in my understanding if i'm incorrect here.

josepharhar replied:
> A span should be allowed for use for the elements that allow phrasing content as part of their content model - and by just referencing the phrasing content allowance, the spec doesn't need to also call out spans then. Otherwise, a div should be sufficient for elements like select and optgroup.

So we should remove span from optgroup and select then, and reserve it for options?

> And for most of the cases where div is being allowed, couldn't it be identified as having a transparent content model?

Yeah I don't know why the content model of div isn't transparent (and span for that matter), and it looks like it already has a carve out for the dl element. I suppose we could make it transparent but I don't know what consequences that might have on other situations and I'd rather try it as a separate thing.

scottaohara commented:
This PR removes some important(?) context about option use within / outside of a datalist. 
for instance, without mention of datalist, the second condition in the option content model doesn't make sense / doesn't appear to do anything with the text inside of the option:
>If the element has a [label](https://whatpr.org/html/10586/form-elements.html#attr-option-label) attribute but no [value](https://whatpr.org/html/10586/form-elements.html#attr-option-value) attribute: [Text](https://whatpr.org/html/10586/dom.html#text-content).

per the above, if i have `<option label=foo>test</option>` with the current select, all browsers use "foo" as the visible text. But, per the current spec, this makes more sense because other conditions are for use of option as not a child of a datalist.  If the option is a child of a datalist, that example renders "test" as the primary label, and "foo" as a sort of sub-text (if rendered at all)... and at least chromium browsers expose that subtext as part of the option...

but again, that only works within a datalist, which would be good to point out (unless we want to do similar for options within a select - which could be useful.....)

Additionally, testing with the current chromium canary build, following the first content model condition results in no visible label for an option
>If the element has a [label](https://whatpr.org/html/10586/form-elements.html#attr-option-label) attribute and a [value](https://whatpr.org/html/10586/form-elements.html#attr-option-value) attribute: [Nothing](https://whatpr.org/html/10586/dom.html#concept-content-nothing).

so `<option value=foo label=test>` renders as an empty option with the customizable select, but renders "foo" as a visible label with the current select.

if the label attribute is not being used in the customizable select, i think this needs to be clearly stated.  (checking/rechecking the definition for option and the label attribute, i don't see this called out?)

but just as importantly, it should also be clear whether new content model allowances are only allowed within options of a styleable select, or if they are allowed for options in a datalist.  the current PR reads as if this is a change for everywhere, and i'm not sure if that's really the intent at this point?

josepharhar commented on source:
```diff
@@ -12975,6 +12975,74 @@ https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C%21DOCTYPE%20HTML%3E%
+  <ul class="brief category-list">
+   <li><span>text content</span></li>
+
+   <li><span>phrasing content</span> except for <span>interactive content</span></li>
```
Thanks! This makes sense. How about this:

Option element inner content elements are div and phrasing content, but there must be no descendants which are datalist, object, or interactive content, and no descendants may have the tabindex attribute specified.

scottaohara replied:
i think that looks good under the premise that option, datalist and object (behaving as an iframe - not sure how to word that) arent just added to the list of interactive elements.

so either what you have, or if those elements are listed under interactive content (because i don't know why they wouldn't be) it could read:
>Option element inner content elements are div and phrasing content, but there must be no descendants which are interactive content, and no descendants may have the tabindex attribute specified.

josepharhar replied:
Ok I adopted the text I proposed and filed a separate issue to discuss interactive elements. If that is addressed, then I can adopt the text you suggested: https://github.com/whatwg/html/issues/10755

josepharhar commented:
> This PR removes some important(?) context about option use within / outside of a datalist.

I re-added the datalist cases and added the new content model under the case where there is no label attribute and there is no parent datalist.

> Additionally, testing with the current chromium canary build, following the first content model condition results in no visible label for an option

This has been fixed

mtrootyy commented on source:
```diff
@@ -24529,10 +24585,14 @@ wormhole connection.&lt;/mark>&lt;/p></code></pre>
    <dd><span>Phrasing content</span>.</dd>
    <dd><span>Palpable content</span>.</dd>
+   <dd><span>Select element inner content elements</span>.</dd>
+   <dd><span>Optgroup element inner content elements</span>.</dd>
```
span element is not "Select element inner content elements" and "Optgroup element inner content elements"

josepharhar replied:
thanks, removed

mtrootyy commented on source:
```diff
@@ -42520,8 +42581,9 @@ interface <dfn interface>HTMLAreaElement</dfn> : <span>HTMLElement</span> {
-  <span>phrasing content</span>, <span>flow content</span>, and <span>palpable content</span>
-  categories for the purposes of the content models in this specification.</p>
+  <span>phrasing content</span>, <span>flow content</span>, <span>palpable content</span>, and
+  <span>option element inner content elements</span> categories for the purposes of the content
```
As long as it is described that it is "phrasing content", it is not necessary to described that it is "Option element inner content elements".
The same is true for "span", "img", "noscript".

josepharhar replied:
Done, thanks!

mtrootyy commented on source:
```diff
@@ -24529,10 +24585,14 @@ wormhole connection.&lt;/mark>&lt;/p></code></pre>
    <dd>Where <span>phrasing content</span> is expected.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd><span>Phrasing content</span>.</dd>
+   <dd>If the element is a descendant of an <code>option</code> element: Zero or more <span>option element inner content elements</span>.</dd>
```
Is this sentence necessary?
If it is necessary here, then it seems necessary for "Content model:" of all of the other element of "Phrasing content" as as well.

josepharhar replied:
I'd like to believe that we don't have to patch the content model in so many ways in so many places in order to define what can go in an option element, so yeah I'm in favor of reverting this change. I likely added it in response to other feedback in this PR though so I'd like to find that first.

zcorpan replied:
It's necessary if we want to have different content models in `select` vs outside.

josepharhar replied:
I think things actually make sense as-is now that we have option element inner content elements defined as phrasing content and `<div>`.

annevk reviewed (commented):
OP says this resolves https://github.com/whatwg/html/issues/10317#issuecomment-2317442832 (among other things) but I don't see that aspect defined here. Where is that defined?

mtrootyy commented on source:
```diff
@@ -54234,17 +54297,18 @@ interface <dfn interface>HTMLOptGroupElement</dfn> : <span>HTMLElement</span> {
-   <span>inter-element whitespace</span>.</dd>
+   child of a <code>datalist</code> element: Zero or more <span>option element inner content
+   elements</span>.</dd>
    <dd>If the element has no <code data-x="attr-option-label">label</code> attribute and is a child
```
To deal with the following cases, the "child" should be changed to "descendant".

```
<label>
 Animal:
 <input name=animal list=animals>
</label>
<datalist id=animals>
 <label>
  or select from the list:
  <select name=animal>
   <option value="">
   <option>Cat
   <option>Dog
  </select>
 </label>
</datalist>
```

josepharhar replied:
done

josepharhar commented:
> work be done to not render both the text from the label attribute AND the legend - which is what's happening now and seems rather rubbish - especially if someone has to write duplicate / close to duplicate content just so they can put an image into their optgroup label

The text in the legend element replaces the text in the label attribute for both rendering and what's exposed to the accessibility tree. The fact that both the label attribute gets rendered when you have a legend is a bug, I'll fix it here: https://issues.chromium.org/issues/378601807

scottaohara commented:
> > work be done to not render both the text from the label attribute AND the legend - which is what's happening now and seems rather rubbish - especially if someone has to write duplicate / close to duplicate content just so they can put an image into their optgroup label
> 
> The text in the legend element replaces the text in the label attribute for both rendering and what's exposed to the accessibility tree. The fact that both the label attribute gets rendered when you have a legend is a bug, I'll fix it here: https://issues.chromium.org/issues/378601807

good to know.  i need to revise the HTML AAM PR for this now, since i was attempting to account for both being rendered until your comment.

josepharhar commented:
> OP says this resolves #10317 (comment) (among other things) but I don't see that aspect defined here. Where is that defined?

I replied on that comment in the issue thread

mtrootyy commented on source:
```diff
@@ -54234,17 +54292,18 @@ interface <dfn interface>HTMLOptGroupElement</dfn> : <span>HTMLElement</span> {
-   <span>inter-element whitespace</span>.</dd>
+   descendant of a <code>datalist</code> element: Zero or more <span>option element inner content
+   elements</span>.</dd>
    <dd>If the element has no <code data-x="attr-option-label">label</code> attribute and is a child
```
child → descendant

josepharhar replied:
Do you mean "descendant -> child"? It currently says descendant.

mtrootyy replied:
```
54248 54307 <dd>If the element has no <code data-x="attr-option-label">label</code> attribute and is a child
````
↓
```
54248 54307 <dd>If the element has no <code data-x="attr-option-label">label</code> attribute and is a descendant
````

zcorpan commented on source:
```diff
@@ -12927,6 +12927,55 @@ https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C%21DOCTYPE%20HTML%3E%
+
+  <p><dfn>Option element inner content elements</dfn> are <code>div</code> and <span>phrasing
+  content</span>, but there must be no descendants which are <code>datalist</code>,
+  <code>object</code>, or <span>interactive content</span>, and no descendants may have the <code
```
"no descendants may" is not a "must"-level requirement. Since this is a definition, I would remove RFC 2119 keywords here, and let the caller use an appropriate keyword.

josepharhar replied:
Agreed, removed

zcorpan commented on source:
```diff
@@ -54328,11 +54386,11 @@ interface <dfn interface>HTMLDataListElement</dfn> : <span>HTMLElement</span> {
+   <dd>As a descendant of a <code>select</code> element.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd>Zero or more <code>option</code> and <span data-x="script-supporting elements">script-supporting</span> elements.</dd>
+   <dd>Zero or one <code>legend</code> element followed by zero or more <span>optgroup element inner content elements</span>.</dd>
```
Should the Rendering section be updated for `legend` in `optgroup` in this PR?

> An [optgroup](https://html.spec.whatwg.org/#the-optgroup-element) element is expected to be rendered by displaying the element's [label](https://html.spec.whatwg.org/#attr-optgroup-label) attribute.

https://html.spec.whatwg.org/#the-select-element-2

josepharhar replied:
Done, I also modified the optgroup label attribute and added an algorithm in there for getting the label from either the legend child or the label attribute. How does it look?

josepharhar commented on source:
```diff
@@ -56171,8 +56265,13 @@ interface <dfn interface>HTMLFieldSetElement</dfn> : <span>HTMLElement</span> {
+   <dd>As the <span>first child</span> of an <code>optgroup</code> element.</dd>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd><span>Phrasing content</span>, optionally intermixed with <span>heading content</span>.</dd>
+   <dd>If the element is the <span>first child</span> of a <code>fieldset</code> element:
+   <span>Phrasing content</span>, optionally intermixed with <span>heading content</span>.</dd>
+   <dd>If the element is the <span>first child</span> of an <code>optgroup</code> element:
+   <span>Phrasing content</span>, but there must be no <span>interactive content</span> and no
+   descendant with the <code data-x="attr-tabindex">tabindex</code> attribute.</dd>
```
done

zcorpan commented on source:
```diff
@@ -56267,11 +56267,11 @@ interface <dfn interface>HTMLFieldSetElement</dfn> : <span>HTMLElement</span> {
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd>If the element is the <span>first child</span> of a <code>fieldset</code> element:
-   <span>Phrasing content</span>, optionally intermixed with <span>heading content</span>.</dd>
    <dd>If the element is the <span>first child</span> of an <code>optgroup</code> element:
```
Replace "the first child" with "a child" here.

josepharhar replied:
Done

mtrootyy commented on source:
```diff
@@ -54425,19 +54483,20 @@ interface <dfn interface>HTMLOptGroupElement</dfn> : <span>HTMLElement</span> {
-   <dd>If the element has no <code data-x="attr-option-label">label</code> attribute and is a child
-   of a <code>datalist</code> element: <span data-x="text content">Text</span>.</dd>
+   descendant of a <code>datalist</code> element: Zero or more <span>option element inner content
+   elements</span>.</dd>
```
```
<select>
<option><em>Text1</em></option>
<option><a href="xx.html">Text2</a></option>
<option><em><a href="xx.html">Text3</a></em></option>
</select>
```

The `<em>Text1</em>` is conformance.
The `<a href="xx.html">Text2</a>` is unconformance.
Well then, the `<em><a href="xx.html">Text3</a></em>` is conformance?

josepharhar replied:
Good point, `<a>` should never be allowed in `<option>`s. 

In this commit I removed the word "descendant" while removing rfc keywords, so I guess I'll just put it back in? https://github.com/whatwg/html/pull/10586/commits/77d2782075007072f6b983a5d570f2592d27c961

mtrootyy replied:
> In this commit I removed the word "descendant" while removing rfc keywords, so I guess I'll just put it back in? https://github.com/whatwg/html/commit/77d2782075007072f6b983a5d570f2592d27c961

Yes, that's right.

josepharhar replied:
Done

mtrootyy commented on source:
```diff
@@ -54374,9 +54432,9 @@ interface <dfn interface>HTMLOptGroupElement</dfn> : <span>HTMLElement</span> {
-  interface. <span w-nodev>User agents should use this attribute's value when labeling the group of
-  <code>option</code> elements in a <code>select</code> element.</span></p>
+  attribute must be specified if the <code>optgroup</code> has no child <code>legend</code> element,
+  and must not be specified if the <code>optgroup</code> has a child <code>legend</code>
```
The label attribute of optgroup element should be "must be specified", to ensure backwards compatible for legacy web browsers that doesn't support legend element in optgroup element.

annevk replied:
Yeah I don't think this is a good idea. We design the conformance criteria for the long term and for the long term requiring `label` would be incredibly annoying. A validator might emit a warning nevertheless to alert web developers about such transitional cases.

cc @sideshowbarker

mtrootyy replied:
Even if it is not "must be specified", at least, specifying of label attribute shouldn't be prohibited, to considerate backwards compatible for legacy web browsers that doesn't support legend element in optgroup element.

sideshowbarker replied:
> We design the conformance criteria for the long term and for the long term requiring `label` would be incredibly annoying. A validator might emit a warning nevertheless to alert web developers about such transitional cases.

Yeah — though for cases where we do have the checker emitting warnings, it’s preferred that there’s some corresponding warning prose in the spec itself.

And incidentally, due to inertia or maintainer laziness or whatever, such warnings that we add to the checker “transitionally” tend to end up hanging around  in the checker much longer past the time when they might have ever actually been useful — until the point where somebody finally takes time to file an issue asking “why in the world in the checker still emitting a warning about this?” or whatever…

Given that, my preference is that we only add new ones if we think they’re really going to help a significant number of web developers avoid some given mistake/problem.

josepharhar replied:
> Even if it is not "must be specified", at least, specifying of label attribute shouldn't be prohibited, to considerate backwards compatible for legacy web browsers that doesn't support legend element in optgroup element.

Ok, I removed the prohibition of the label attribute in the case that there is a legend element

scottaohara commented:
> > This could use some examples. Or are we adding those as part of another PR?
> 
> I started a follow up here: https://github.com/whatwg/html/compare/main...josepharhar:html:selectexamples?expand=1

thanks @josepharhar - give me a shout when these are finished, and i can help review and/or add on as need be.

josepharhar commented:
In order to resolve the remaining two open comment threads, I added notes about the content model:
- https://github.com/whatwg/html/pull/10586/commits/473ca72a08e0e6fb52af1373cb8a66b0ca51e265
- https://github.com/whatwg/html/pull/10586#discussion_r1798109306
- https://github.com/whatwg/html/pull/10586#discussion_r1807576729

I originally had notes about the content model which were placed directly after the content model itself, but I was told not to do that and to put the notes where the prose about the content model is. I'm not sure where such prose exists, so I just added the notes to the very end of the option element and select element sections: https://github.com/whatwg/html/pull/10586#discussion_r1761323296

domfarolino reviewed (approved):
It looks like everything has been addressed with this PR for a bit now. The contents look good to me, thanks for resolving the last two bits in https://github.com/whatwg/html/pull/10586#issuecomment-2722397769 @josepharhar and adding more detail to the commit message in the OP.

I'll plan on merging this towards the end of next week, provided the one nit I found gets resolved and there are no further substantial reviews.

josepharhar commented on source:
```diff
@@ -54707,6 +54796,54 @@ interface <dfn interface>HTMLOptionElement</dfn> : <span>HTMLElement</span> {
+       <code data-x="">" "</code>, the value of <span>descendant</span>'s <code
+       data-x="attr-img-alt">alt</code> attribute, and <code data-x="">" "</code>.</p></li>
+
+       <li><p><span>Continue</span> skipping all <span data-x="descendant">>descendants</span> of
```
fixed, thanks

mtrootyy commented on source:
```diff
@@ -24490,7 +24549,9 @@ wormhole connection.&lt;/mark>&lt;/p></code></pre>
    <dt><span data-x="concept-element-content-model">Content model</span>:</dt>
-   <dd><span>Phrasing content</span>.</dd>
+   <dd>If the element is a descendant of an <code>option</code> element: Zero or more
+   <span><code>option</code> element inner content elements</span>.</dd>
```
Is the following conformance?

```
<select>
<option><span><div>option1</div></span></option>
<option><b><span><div>option2</div></span></b></option>
</select>

```

josepharhar replied:
Yes, that code doesn't pose any accessibility issues that I'm aware of, and we want to allow those elements inside of option elements.

mtrootyy replied:
If so, some addition note would be needed in the section of "15 Rendering".
Because it should be clarify how to render the element of "display:inline" which has div element in desendant.

josepharhar replied:
Ok, I see that without this change, the content model of span doesn't allow divs. I updated this text to exclude divs.

annevk commented on source:
```diff
@@ -12939,6 +12939,55 @@ https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C%21DOCTYPE%20HTML%3E%
+  <p><dfn><code>select</code> element inner content elements</dfn> are the elements which are
+  allowed as descendants of <code>select</code> elements.</p>
+
+  <p>The following are <span><code>select</code> element inner content elements</span>:</p>
```
https://github.com/whatwg/html/pull/10633 seems to expect that the `select` element can have a `button` element child. Shouldn't that be described here? Is that PR going to be changed and rebased after this lands? It would be good to see the overall plan here.

josepharhar replied:
This PR lets the select element have a button child, and the `<selectedcontent>` PR lets selectedcontent be put inside the button element. Besides merging this one before the selectedcontent PR, what changes do you think should be made?

annevk replied:
I don't see `button` listed?

annevk commented on source:
```diff
@@ -54603,6 +54668,13 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
   &hellip;</code></pre>
   </div>
 
+  <div class="note">
+   <p>The first child <code>button</code> element as allowed by the content model of
+   <code>select</code> is not a submit button. It is used to replace the in-page rendering of the
+   <code>select</code> element. Its form submission behavior is prevented because it has inert set
+   in the UA stylesheet.</p>
+  </div>
```
This presupposes a feature that doesn't exist. Can we solve this differently? Perhaps we can forcibly make this not a submit button and also reflect that in the `type` getter?

josepharhar replied:
Yes, we could easily solve this differently by just saying that the first child `<button>` is inert instead of putting that in the UA stylesheet.

Inertness helps more than just making the button not a submit button, such as ensuring accessibility by not allowing nested interactive content. I could also forcibly make this not a submit button, but in practice that doesn't matter if the button is inert.

What do you think?

annevk replied:
Are you sure it doesn't matter? Wouldn't it still be chosen as the default submit button for a form when the end user presses enter or some such if it happens to be before another submit button?

josepharhar replied:
I have a fix for it being the default submit button in the base appearance PR here: https://github.com/whatwg/html/pull/10629/files#diff-41cf6794ba4200b839c53531555f0f3998df4cbb01a4d5cb0b94e3ca5e23947dR60309

I suppose that would be better handled in this PR, so I moved it from there to here and tweaked it a bit to not talk about the button slot.

I also made the first child button of a select inert and removed it from the CSS PR

annevk replied:
But why fix this problem in that way instead of making it not be a submit button? That way we don't have to change any of the "submit button" branches.

josepharhar replied:
Makes sense, I pushed a commit to make it not a submit button.
